pkg/aks-engine/helpers/helpers.go

```diff
@@ -0,0 +1,108 @@
+func GetHomeDir() string {
+	if runtime.GOOS == "windows" {
+		home := os.Getenv("HOMEDRIVE") + os.Getenv("HOMEPATH")
+		if home == "" {
+			home = os.Getenv("USERPROFILE")
+		}
+		return home
+	}
+	return os.Getenv("HOME")
+}
+
+// CreateSSH creates an SSH key pair.
+func CreateSSH(rg io.Reader, s *i18n.Translator) (privateKey *rsa.PrivateKey, publicKeyString string, err error) {
+	privateKey, err = rsa.GenerateKey(rg, SSHKeySize)
+	if err != nil {
+		return nil, "", s.Errorf("failed to generate private key for ssh: %q", err)
+	}
+
+	publicKey := privateKey.PublicKey
+	sshPublicKey, err := ssh.NewPublicKey(&publicKey)
+	if err != nil {
+		return nil, "", s.Errorf("failed to create openssh public key string: %q", err)
+	}
+	authorizedKeyBytes := ssh.MarshalAuthorizedKey(sshPublicKey)
+	authorizedKey := string(authorizedKeyBytes)
+
+	return privateKey, authorizedKey, nil
+}
+
+// JSONMarshalIndent marshals formatted JSON w/ optional SetEscapeHTML
+func JSONMarshalIndent(content interface{}, prefix, indent string, escape bool) ([]byte, error) {
+	b, err := JSONMarshal(content, escape)
+	if err != nil {
+		return nil, err
+	}
+
+	var bufIndent bytes.Buffer
+	if err := json.Indent(&bufIndent, b, prefix, indent); err != nil {
+		return nil, err
+	}
+
+	return bufIndent.Bytes(), nil
+}
+
+// JSONMarshal marshals JSON w/ optional SetEscapeHTML
+func JSONMarshal(content interface{}, escape bool) ([]byte, error) {
+	var buf bytes.Buffer
+	enc := json.NewEncoder(&buf)
+	enc.SetEscapeHTML(escape)
+	if err := enc.Encode(content); err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+// GetCloudTargetEnv determines and returns whether the region is a sovereign cloud which
+// have their own data compliance regulations (China/Germany/USGov) or standard
+// Azure public cloud
+func GetCloudTargetEnv(location string) string {
+	loc := strings.ToLower(strings.Join(strings.Fields(location), ""))
+	switch {
+	case loc == "chinaeast" || loc == "chinanorth" || loc == "chinaeast2" || loc == "chinanorth2":
+		return "AzureChinaCloud"
+	case loc == "germanynortheast" || loc == "germanycentral":
+		return "AzureGermanCloud"
+	case strings.HasPrefix(loc, "usgov") || strings.HasPrefix(loc, "usdod"):
+		return "AzureUSGovernmentCloud"
+	default:
+		return "AzurePublicCloud"
+	}
+}
+
+// GetTargetEnv determines and returns whether the region is a sovereign cloud which
+// have their own data compliance regulations (China/Germany/USGov) or standard
+// Azure public cloud
+// CustomCloudName is name of environment if customCloudProfile is provided, it will be empty string if customCloudProfile is empty.
+// Because customCloudProfile is empty for deployment for AzurePublicCloud, AzureChinaCloud,AzureGermanCloud,AzureUSGovernmentCloud,
+// The customCloudName value will be empty string for those clouds
+func GetTargetEnv(location, customCloudName string) string {
+	switch {
+	case customCloudName != "" && strings.EqualFold(customCloudName, "AzureStackCloud"):
```

Comment:
we don't need this

Reply:
Used in pkg/agent/datamodel/types.go:
targetEnv := helpers.GetTargetEnv(cs.Location, cs.Properties.GetCustomCloudName())

But maybe we can remove that code in pkg/agent/datamodel/types.go as well?